ACMILabs/playlist-label
# Update/53 frontend to reconnect from network failure

Resolves #47, #53 & ACMILabs/xos#2048

Playlist Label keeps trying to reconnect to the MQTT broker after recovering from a network failure.

### Acceptance Criteria
- [x] Keep trying to re-connect when the MQTT client fails to connect.
- [x] When Playlist is booted and can't access XOS within 5 secs then use cached config and content.

### Relevant design files
* None

### Run manual tests
**Python tests**
1. First terminal window: `cd development` and `docker-compose up --build`
1. Second terminal window: `docker exec -it playlistlabel make linttest`

**Javascript tests**
1. First terminal window : `cd testing` and `docker-compose up --build`
1. Second terminal window : `docker exec -it javascripttests make linttestjs`

### Testing instructions
1. Start the development environment: `docker-compose up --build`
1. With a media player successfully updating the playlist label, disconnect from the VPN and see the updating stop
1. Re-connect to the VPN and see that the frontend and status bar starts updating again
1. Now try starting the Playlist Label without the VPN connected and after 5 seconds or so see that the request to XOS times out and it loads from the cached file: `docker-compose up`
1. Delete any cached `playlist_XX.json` files at the root of the directory, disconnect your VPN and restart the playlist player: `docker-compose up`
1. Note that you see the `Playlist not found` screen with suggestions
1. See this on a real Raspberry Pi by adding it to this Balena application: [s__playlist-label-pi-4](https://dashboard.balena-cloud.com/apps/1529802/devices)
1. Test that it loads the empty playlist screen starting without any internet access
1. Test that it loads a cached playlist without any internet access after successfully caching one
1. Test that it recovers gracefully when internet is added and starts updating again

![Screen Shot 2020-06-16 at 3 23 41 pm](https://user-images.githubusercontent.com/314298/84736902-f752ed00-afe5-11ea-8a0e-9a928c9bf07d.png)

### DoD
For requester to complete:
